### PR TITLE
F#src 2601 adding watchdog

### PR DESCRIPTION
**Pull request by mikramarc**

### CODE REVIEW

File: sr_utilities_common/README.md [hunk 1]
```diff
@@ -26,4 +26,63 @@ if __name__ == '__main__':
 
 ```
 
-The above code will be executing an infinite loop until shutdown for the node is initiated. Then, before the node dies, the `clean_up` method will be executed. 
\ No newline at end of file
+The above code will be executing an infinite loop until shutdown for the node is initiated. Then, before the node dies, the `clean_up` method will be executed. 
+
+## Watchdog
+A watchdog class that continuously runs specified check methods and informs user about any undesired behaviours.
```
Comment (ethanfowler) [inline sr_utilities_common/README.md]: users/the user

File: sr_utilities_common/README.md [hunk 2]
```diff
@@ -26,4 +26,63 @@ if __name__ == '__main__':
...
+    error_checks_list = ['mock_check_if_arm_running']
+    warning_checks_list = ['mock_check_robot_clear_from_collision']
+
+    teleop_watchdog = SrWatchdog(test_class, error_checks_list, warning_checks_list)
+    shutdown_handler = ShutdownHandler(teleop_watchdog, 'clean_up()')
+    teleop_watchdog.run()
+    rospy.spin()
+```
+
+All the check methods are defined in a separate class. There are two types of checks supported. Error checks, which will change general status to `error` and throw an errro message, and warning checks, which do not affect the reported status but inform user that something has happened. In order to classify a check as error or warning check, their method name needs to be put in a proper list that is further passed to the `SrWatchdog` class constructor, together with the checks class object.
```
Comment (ethanfowler) [inline sr_utilities_common/README.md]: throw an error message
users/the user

File: sr_utilities_common/README.md [hunk 3]
```diff
@@ -26,4 +26,63 @@ if __name__ == '__main__':
...
+
+The check methods are required to return either `True` (for passsed check) or `False` (for failed check) values. Alternatively, in case of failure, error checks can return a tuple of `(False, <error message>)` where `<error message>` is additional, user-defined error message.
```
Comment (ethanfowler) [inline sr_utilities_common/README.md]: passsed -> passed

File: sr_utilities_common/launch/mock_watchdog.launch [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<launch>
+    <node name="mock_watchdog" pkg="sr_utilities_common" type="watchdog.py" output="screen" launch-prefix="xterm -e" />
+</launch>
```
Comment (ethanfowler) [inline sr_utilities_common/launch/mock_watchdog.launch]: Newline at end of file

File: sr_utilities_common/scripts/sr_utilities_common/watchdog.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+                self.demo_status = Status.OK
+
+    def run_error_checks(self):
+        self.run_status_checks('err')
+
+    def run_warning_checks(self):
+        self.run_status_checks('warn')
+
+    def get_cpu_usage(self):
+            self.cpu_usage = psutil.cpu_percent(interval=1, percpu=True)
```
Comment (ethanfowler) [inline sr_utilities_common/scripts/sr_utilities_common/watchdog.py]: Maybe want percpu=False?

File: sr_utilities_common/README.md (hunk 1 above)
Comment (mikramarc) [reply sr_utilities_common/README.md]: fixed in 59ba8a3

File: sr_utilities_common/README.md (hunk 2 above)
Comment (mikramarc) [reply sr_utilities_common/README.md]: fixed in 59ba8a3

File: sr_utilities_common/README.md (hunk 3 above)
Comment (mikramarc) [reply sr_utilities_common/README.md]: fixed in 59ba8a3

File: sr_utilities_common/launch/mock_watchdog.launch (hunk 4 above)
Comment (mikramarc) [reply sr_utilities_common/launch/mock_watchdog.launch]: fixed in 59ba8a3

File: sr_utilities_common/scripts/sr_utilities_common/watchdog.py (hunk 5 above)
Comment (mikramarc) [reply sr_utilities_common/scripts/sr_utilities_common/watchdog.py]: fixed in 758db1d

**Review by ethanfowler — APPROVED** [review]

**Pull request merged**